### PR TITLE
john-jumbo: add upstream ARM64/M1 support

### PR DESCRIPTION
**Pull request by DomT4**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [ ] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

Strict audit failure:
```
  * Formula john-jumbo contains deprecated SPDX licenses: ["GPL-2.0"].
    You may need to add `-only` or `-or-later` for GNU licenses (e.g. `GPL`, `LGPL`, `AGPL`, `GFDL`).
    For a list of valid licenses check: https://spdx.org/licenses/
Error: 1 problem in 1 formula detected
```

Been about 3 years since my last PR, so, bear with me on how the heck PRs work these days 😸.

### CODE REVIEW

**Comment by DomT4** [comment]
```
==> Summary
🍺  /opt/homebrew/Cellar/john-jumbo/1.9.0: 447 files, 75.9MB, built in 1 minute 26 seconds

~> file /opt/homebrew/Cellar/john-jumbo/1.9.0/bin/john
/opt/homebrew/Cellar/john-jumbo/1.9.0/bin/john: Mach-O 64-bit executable arm64
```

**Comment by Bo98** [comment]
Makes sense. Probably worth hosting the patch on https://github.com/Homebrew/formula-patches instead though.

File: Formula/john-jumbo.rb [hunk 1]
```diff
@@ -53,12 +53,25 @@ class JohnJumbo < Formula
...
 
+  # Upstream M1/ARM64 Support.
+  # Combined diff of the following four commits, minus the doc changes
+  # that block this formula from using these commits otherwise.
+  # https://github.com/openwall/john/commit/d6c87924b85323b82994ce01724d6e458223fd36
+  # https://github.com/openwall/john/commit/d531f97180a6e5ae52e21db177727a17a76bd2b4
+  # https://github.com/openwall/john/commit/c9825e688d1fb9fdd8942ceb0a6b4457b0f9f9b4
+  # https://github.com/openwall/john/commit/716279addd5a0870620fac8a6e944916b2228cc2
+  patch do
+    url "https://gist.githubusercontent.com/DomT4/80e8e8f5477c91c124b727df2ace55c7/raw/022a60fbae8ac1f913c719da6c7152620beaf511/john_jumbo_m1.diff"
```
Comment (SMillerDev) [inline Formula/john-jumbo.rb:64]: Can we rehost this in homebrew/Homebrew-patches?

Comment (SMillerDev) [reply Formula/john-jumbo.rb:64]: I should read other comments first 😅

Comment (DomT4) [reply Formula/john-jumbo.rb:64]: Heh 😄. I'll move the patch over later today & update the PR, yeah.

**Comment by DomT4** [comment]
https://github.com/Homebrew/formula-patches/pull/361

**Comment by DomT4** [comment]
Changes pushed.

File: Formula/john-jumbo.rb [hunk 2]
```diff
@@ -53,12 +53,25 @@ class JohnJumbo < Formula
...
+  end
+
   def install
     ENV.append "CFLAGS", "-DJOHN_SYSTEMWIDE=1"
     ENV.append "CFLAGS", "-DJOHN_SYSTEMWIDE_EXEC='\"#{share}/john\"'"
     ENV.append "CFLAGS", "-DJOHN_SYSTEMWIDE_HOME='\"#{share}/john\"'"
 
-    ENV.append "CFLAGS", "-mno-sse4.1" unless MacOS.version.requires_sse4?
+    # Apple's M1 chip has no support for SSE 4.1.
+    ENV.append "CFLAGS", "-mno-sse4.1" if Hardware::CPU.intel? && !MacOS.version.requires_sse4?
```
Comment (SMillerDev) [inline Formula/john-jumbo.rb:74]: wouldn't it need to be added on M1 too? Like:
```suggestion
    # Apple's M1 chip has no support for SSE 4.1.
    ENV.append "CFLAGS", "-mno-sse4.1" unless MacOS.version.requires_sse4? || Hardware::CPU.intel?
```

Comment (Bo98) [reply Formula/john-jumbo.rb:74]: Nah, SSE doesn't exist on ARM so it's impossible for the compiler to even try use it. `-mno-sse4.1` possibly even errors on ARM64.

Comment (SMillerDev) [reply Formula/john-jumbo.rb:74]: so... it's not `-m no-sse4.1` then? I was expecting this to disable the feature

Comment (Bo98) [reply Formula/john-jumbo.rb:74]: SSE flags do nothing on non-x86_64 platforms.

I thought they might error but it seems they warn instead in Clang:

```
argument unused during compilation: '-mno-sse4.1' [-Wunused-command-line-argument]
```

On GCC, it is a hard error on non-x86_64 platforms:

```
aarch64-unknown-linux-gnu-g++: error: unrecognized command-line option '-mno-sse4.1'
```

Comment (DomT4) [reply Formula/john-jumbo.rb:74]: Homebrew itself stops you even trying on ARM/Apple Silicon, which is reasonable but not immediately obvious what's happened to the end user:
```
Error: An exception occurred within a child process:
  RuntimeError: Unexpected architecture: arm64. This only works with Intel architecture.
```

Comment (Bo98) [reply Formula/john-jumbo.rb:74]: Ah, yes. Too busy looking at the flag without realising `MacOS.version.requires_sse4?` errors.

**Review by Bo98 — APPROVED** [review]

**Comment by Bo98** [comment]
I'll merge this and then start a separate arm64-only bottle build, since we don't need to discard old Intel bottles.

**Comment by Bo98** [comment]
https://github.com/Homebrew/homebrew-core/actions/runs/752890982

**Comment by DomT4** [comment]
Thanks @Bo98! 🎆

**Comment by Bo98** [comment]
https://github.com/Homebrew/homebrew-core/commit/9453783634a49923ace2b9d8aa3962f33ac01661 🚀

**Pull request merged**